src/System.Private.Uri/src/System/Uri.cs

```diff
@@ -4672,6 +4672,11 @@ private unsafe Check CheckCanonical(char* str, ref ushort idx, ushort end, char
                         res |= Check.NotIriCanonical;
                     }
                 }
+                else if ((c >= '{' && c <= '}'))
+                {
+                    if (!needsEscaping)
```

Comment:
Is it really necessary to check for false before setting to true?  One place in the code above uses that pattern.  Another place simply sets it to 'true'.  I would change this to just 'needsEscaping = true;'